### PR TITLE
azurerm_virtual_machine_scale_set - prevent `public_ip_address_configuration` from being lost during update

### PR DESCRIPTION
**Pull request by alexsapran**
Why 
----
This commit is fixing #3763 when we perform a simple update on an existing scale-set the definition of the scale-set was changed on the `publicIPAddressConfiguration` field. 

This means that any new instances that would be started from the scale-set would not have public IPs when it would be expected to have public IP.

What
----
* Create acceptance tests to mitigate the issue

Start with creating some test coverage around the issue. 

* Fix the parsing for the public ip configuration

Fix the actual issue that caused the public ip configuration of a scale-set to be removed.

Fixes: #3763

### CODE REVIEW

File: azurerm/resource_arm_virtual_machine_scale_set.go [hunk 1]
```diff
@@ -1331,8 +1331,9 @@ func flattenAzureRmVirtualMachineScaleSetNetworkProfile(profile *compute.Virtual
 						publicIpConfigs := make([]map[string]interface{}, 0, 1)
 						publicIpConfig := make(map[string]interface{})
 						publicIpConfig["name"] = *publicIpInfo.Name
-						publicIpConfig["domain_name_label"] = *publicIpInfo.VirtualMachineScaleSetPublicIPAddressConfigurationProperties.DNSSettings
+						publicIpConfig["domain_name_label"] = *publicIpInfo.VirtualMachineScaleSetPublicIPAddressConfigurationProperties.DNSSettings.DomainNameLabel
```
Comment (kwilczynski) [inline azurerm/resource_arm_virtual_machine_scale_set.go]: Nice catch!

**Comment by kwilczynski** [comment]
Hi @jeffreyCline, the fix is very simple and appears to solve the problem. Do you think you could have a look for us?

File: azurerm/resource_arm_virtual_machine_scale_set.go (hunk 1 above)
Comment (katbyte) [reply azurerm/resource_arm_virtual_machine_scale_set.go]: Indeed! We are going to definitely need to add some.. many nil checks to this as the original code was lacking many. My suggestion is something like this:
```go
if ipCfg := properties.PublicIPAddressConfiguration; ipCfg != nil {
						block := make(map[string]interface{})
						if v := ipCfg.Name; v != nil {
							block["name"] = *v
						}

						if ipProperties := ipCfg.VirtualMachineScaleSetPublicIPAddressConfigurationProperties; ipProperties != nil {
							if dns := ipProperties.DNSSettings; dns != nil {
								if v := dns.DomainNameLabel; v != nil {
									block["domain_name_label"] = *v
								}
							}
							if v := ipProperties.IdleTimeoutInMinutes; v != nil {
								block["idle_timeout"] = *v
							}
						}
						config["public_ip_address_configuration"] = []map[string]interface{}{block}
					}
```

**Comment by kwilczynski** [comment]
Hi @katbyte and @jeffreyCline, thank you so much for the review!

I believe @alexsapran will be able to include the `nil` checks in the morning. Which brings me to a question: do you have more release-worthy features lined up to warrant a new version release, so that we could upgrade to an official upstream release ourselves? If no, then perhaps enough bug fixes lined up to make a point-release a reality?

**Comment by katbyte** [comment]
@kwilczynski,

given that we just released last friday and there are only a handful unreleased so most likely release date will be next week.

**Comment by alexsapran** [comment]
Hi @jeffreyCline and @katbyte and thank you for taking the time to review my PR. 

I have pushed 2 new commits ea555d0 and 1498394. The first is just to follow the pattern of the nil checks and the second is to follow your recommendation around the additional Nil checks that you requested.

Please let me know what do you think and if you are still +1

**Comment by kwilczynski** [comment]
Hi @katbyte, thank you so much for your help and for cutting the release. Much appreciated!

**Pull request merged**